src/git.rs

```diff
@@ -79,40 +115,181 @@ impl Git for LiveGit {
-    use super::*;
-    
-    pub struct MockGit;
-    impl Git for MockGit {
-        fn new() -> Result<Self, Box<dyn Error>> {
-            Ok(MockGit)
+    fn log(&self) -> Result<(), Box<dyn Error>> {
+        let mut revwalk = self.repo.revwalk()?;
+        revwalk.push_head()?;
+        for id in revwalk {
+            let id = id?;
+            let commit = self.repo.find_commit(id)?;
+            Self::debug_commit(&commit);
+        }
+        Ok(())
+    }
+
+    fn find_commit(&self, commit_id: Oid) -> Result<Commit<'_>, Box<dyn Error>> {
+        let new_commit = self.repo.find_commit(commit_id)?;
+        Ok(new_commit)                
+    }
+
+    fn find_commit_fingerprints(&self, team_fingerprint_file: &str, commits: &Vec<Commit<'_>>) -> Result<HashSet<String>, Box<dyn Error>> {
+
+        let team_fingerprints = Fingerprint::read_fingerprints::<LiveGit>(self, team_fingerprint_file)?;
+
+        let mut commit_fingerprints = HashSet::new();
+
+        for commit in commits.iter() {
+            let author = commit.author();
+            let commit_email = match author.email() {
+                Some(e) => e,
+                None => return Err(Box::new(CapnError::new(format!("Email on commit not found"))))
+            };
+            let fingerprint = team_fingerprints.iter().find(|f| f.email == commit_email);
+            match fingerprint{
+                Some(f) => commit_fingerprints.insert(f.id.to_string()),
+                None => return Err(Box::new(CapnError::new(format!("Team fingerprint not found"))))
+            };
+        }
+        Ok(commit_fingerprints)
+    }
+
+    fn pushed(&self, commit_id: Oid) -> Result<bool, Box<dyn Error>> {
+        debug!("Check if commit {} has already been pushed", commit_id);
+
+        let repo_path = self.repo.path();
+        debug!("Repo path {:?}", repo_path);
+        let result = Command::new("git")
+            .current_dir(repo_path)
+            .arg("branch")
+            .arg("--contains")
+            .arg(commit_id.to_string())
+            .output()?;
+        debug!("RESULT {:?}", result);
+        if !result.status.success() {
+            return Err(Box::new(CapnError::new(format!("Call to git branch contains failed for commit {} with status {}", commit_id,result.status))));
+        }
+
+        let output = String::from_utf8(result.stdout)?;
+        match output.trim() {
+            "" => return Ok(false),
+            _ => return Ok(true)
+        };                
+    }
+
+    fn find_commits(&self, from_id: Oid, to_id: Oid) -> Result<Vec<Commit<'_>>, Box<dyn Error>> {
+        debug!("Find commits between {} to {}", from_id, to_id);
+
+        let mut v = Vec::new();
+        let new_commit = self.repo.find_commit(to_id)?;
+        let mut current_id = to_id;
+        v.push(new_commit);
+        while current_id != from_id {
+            let current_commit = self.repo.find_commit(current_id)?;
+            for parent in current_commit.parents() {
```

Comment:
Aren't we only allowing merge commits on the target branch?
Having more than one here would be a failure condition and should probably break out.

Reply:
I don't think that this algorithm would explore the tree exhaustively anyway.  It looks like you would always search which ever parent was last.  Or am I missing something?